src/eth1spec/spec.py

```diff
@@ -330,6 +342,126 @@ def signing_hash(tx: Transaction) -> Hash32:
+    header : `eth1spec.eth_types.Header`
+        Header of interest.
+
+    Returns
+    -------
+    hash : `eth1spec.eth_types.Hash32`
+        Hash of the Header.
+    """
+    return crypto.keccak256(
+        rlp.encode(
+            (
+                header.parent_hash,
+                header.ommers_hash,
+                header.coinbase,
+                header.state_root,
+                header.transactions_root,
+                header.receipt_root,
+                header.bloom,
+                header.difficulty,
+                header.number,
+                header.gas_limit,
+                header.gas_used,
+                header.timestamp,
+                header.extra_data,
+                header.mix_digest,
+                header.nonce,
+            )
+        )
+    )
+
+
+def get_block_header_by_hash(hash: Hash32, chain: BlockChain) -> Header:
+    """
+    Fetches the block header by searching its hash.
+
+    Parameters
+    ----------
+    hash : `eth1spec.eth_types.Hash32`
+        hash of the header of interest.
+
+    chain : `eth1spec.eth_types.BlockChain`
+        History and current state.
+
+    Returns
+    -------
+    Header : `eth1spec.eth_types.Header`
+        Block header found by its hash.
+    """
+    for block in chain.blocks:
+        if header_hash(block.header) == hash:
+            return block.header
+    else:
+        raise ValueError("Could not find Header with hash {!r}".format(hash))
+
+
+def validate_gas_limit(gas_limit: Uint, parent_gas_limit: Uint) -> bool:
+    """
+    Validates the gas limit for a block
+
+    Parameters
+    ----------
+    gas_limit : `eth1spec.eth_types.Uint`
+        gas limit of the header of interest
+
+    parent_gas_limit : `eth1spec.eth_types.Uint`
+        gas limit of the porent header of interest
+
+    Returns
+    -------
+    verified : `bool`
+        True if header gas limit constraints are satisfied, False otherwise
+    """
+    if gas_limit >= parent_gas_limit + (
+        parent_gas_limit // GAS_LIMIT_ADJUSTMENT_FACTOR
+    ):
+        return False
+    if gas_limit <= parent_gas_limit - (
+        parent_gas_limit // GAS_LIMIT_ADJUSTMENT_FACTOR
+    ):
+        return False
+    if gas_limit < GAS_LIMIT_MINIMUM:
+        return False
+
+    return True
+
+
+def calculate_block_difficulty(header: Header, parent_header: Header) -> Uint:
+    """
+    Computes difficulty of a block using its header and parent header
+    Parameters
+    ----------
+    header: `eth1spec.eth_types.Header`
+        header of interest
+    parent_header: `eth1spec.eth_types.Header`
+        parent header of header of interest
+    Returns
+    ------
+    difficulty: Uint
+        computed difficulty for a block
+    """
+    if parent_header.number == 0:
+        return Uint(131072)
+    elif header.timestamp < parent_header.timestamp:
+        return parent_header.difficulty + parent_header.difficulty // Uint(
+            2048
+        )
+    else:
```

Comment:
Another style choice that extends beyond this repository.  This else block is an *implicit* conditional for `elif header.timestamp > parent_header.timestamp`.  But the condition isn't actually written which makes understanding the logic just a tiny bit harder.  Options for improving this:

```python
# use a comment
if thing == 1:
    ...
elif thing > 1:
    ...
else:  # thing < 1
    ...
    
# explicit invariant
if thing == 1:
    ...
elif thing > 1:
    ...
elif thing < 1:
    ...
else:  # unreachable
    raise Exception("Unreachable")
```

Reply:
I don't think I like either of these styles. Personally, I think most developers understand if/elif/else statements fairly well.

If the conditions ever change (say in an EIP), the comment could get out of sync with the code, and I don't think the tooling we have can actually assert that the `else: # unreachable` is actually unreachable.

Reply:
I don't feel strongly about this.  It's simply a convention that we adopted on our team as a standard convention due to a few times that we've run into bugs due to implicit else clauses that either never get hit *or* that are hard to reason about.  This isn't really either of those.

Regarding tooling:  I don't think this actually applies but here's the best I'm aware of for exhaustiveness checking in python: https://hakibenita.com/python-mypy-exhaustive-checking

Reply:
Oh that's clever! I'll play around with that.